File: rinha-offline/src/logic/Galo.ts
import { calcularMultiplicadorRebirth } from '../store/jogadorStore';
import { GALOS_DB } from '../data/galosDb';
import type { Skill } from '../data/galosDb';
import { EFEITOS_INICIO_TURNO, EFEITOS_DANO_RECEBIDO } from './efeitos';
import { SISTEMA_TIPOS } from '../data/tiposDb';

export type DadosEfeito = {
    nome: string;
    chance: number;
    turnos: number;
};

export class Galo {
    nome: string;
    tipo: string;
    nivel: number;
    xp: number;
    hp_max: number;
    hp_atual: number;
    caminho_imagem: string;
    skills_equipadas: (Skill | null)[];
    rebirths: number;
    efeitos: Record<string, number>;

    constructor(
        nome: string,
        hp_max: number,
        caminho_imagem: string,
        nivel: number = 1,
        xp: number = 0,
        tipo: string = "Normal",
        skills_equipadas: (Skill | null)[] | null = null,
        efeitos: Record<string, number> | null = null,
        rebirths: number = 0
    ) {
        this.nome = nome;
        this.tipo = tipo;
        this.nivel = nivel;
        this.xp = xp;
        this.hp_max = hp_max;
        this.hp_atual = hp_max;
        this.caminho_imagem = caminho_imagem;
        this.skills_equipadas = skills_equipadas ?? [];
        this.efeitos = efeitos ?? {};
        this.rebirths = rebirths;
    }

    obterSkillsDesbloqueadas(): { level: number, skill: Skill }[] {
        if (!GALOS_DB[this.nome]) {
            return [];
        }
        
        const skillsDb = GALOS_DB[this.nome].skills;
        const desbloqueadas: { level: number, skill: Skill }[] = [];
        
        // Em TypeScript, a conversão da chave do objeto para número é necessária para manter a lógica original.
        for (const [lvlStr, skill] of Object.entries(skillsDb)) {
            const lvl = parseInt(lvlStr, 10);
            if (lvl <= this.nivel) {
                desbloqueadas.push({ level: lvl, skill });
            }
        }
        
        return desbloqueadas;
    }

    equiparSkillsBot(): void {
        const desbloqueadas = this.obterSkillsDesbloqueadas();
        desbloqueadas.sort((a, b) => b.level - a.level);
        
        this.skills_equipadas = desbloqueadas.slice(0, 5).map(item => item.skill);
        while (this.skills_equipadas.length < 5) {
            this.skills_equipadas.push(null);
        }
    }

    atacar(): [string, number, DadosEfeito | null] {
        let skillsValidas = this.skills_equipadas.filter(s => s !== null) as Skill[];
        
        if (skillsValidas.length === 0) {
            this.equiparSkillsBot();
            skillsValidas = this.skills_equipadas.filter(s => s !== null) as Skill[];
        }
        
        if (skillsValidas.length === 0) {
            return ["Ataque Básico", 10, null];
        }

        const skill = skillsValidas[Math.floor(Math.random() * skillsValidas.length)];
        const danoTotal = Math.floor(Math.random() * (skill.max - skill.min + 1)) + skill.min;
        
        let dadosEfeito: DadosEfeito | null = null;
        if (skill.efeito) {
            dadosEfeito = {
                nome: skill.efeito,
                chance: skill.chance ?? 100,
                turnos: skill.turnos ?? 1
            };
        }
        
        return [skill.nome, danoTotal, dadosEfeito];
    }

    aplicarEfeito(dadosEfeito: DadosEfeito | null): boolean {
        if (!dadosEfeito) return false;
        
        const roll = Math.floor(Math.random() * 100) + 1;
        if (roll <= dadosEfeito.chance) {
            this.efeitos[dadosEfeito.nome] = dadosEfeito.turnos;
            return true;
        }
        
        return false;
    }

    processarEfeitosInicioTurno(): [string[], boolean] {
        const mensagens: string[] = [];
        let podeAtacar = true;
        const efeitosParaRemover: string[] = [];

        for (const [efeito, turnos] of Object.entries(this.efeitos)) {
            if (turnos > 0) {
                if (EFEITOS_INICIO_TURNO[efeito]) {
                    const [msg, permiteAtacar] = EFEITOS_INICIO_TURNO[efeito](this, efeito);
                    if (msg) mensagens.push(msg);
                    if (!permiteAtacar) podeAtacar = false;
                }

                this.efeitos[efeito] -= 1;
                if (this.efeitos[efeito] <= 0) {
                    efeitosParaRemover.push(efeito);
                }
            }
        }

        for (const efeito of efeitosParaRemover) {
            delete this.efeitos[efeito];
        }

        if (this.hp_atual < 0) {
            this.hp_atual = 0;
        }

        return [mensagens, podeAtacar];
    }

    sofrerDano(danoRecebido: number, tipoAtacante?: string): number {
        let danoReal = Math.max(1, danoRecebido);

        if (tipoAtacante && SISTEMA_TIPOS[tipoAtacante]) {
            const relacao = SISTEMA_TIPOS[tipoAtacante];
            if (relacao.vantagem.includes(this.tipo)) {
                danoReal = Math.round(danoReal * 1.15);
            } else if (relacao.desvantagem.includes(this.tipo)) {
                danoReal = Math.round(danoReal * 0.90);
            }
        }

        const efeitosParaRemover: string[] = [];
        
        // No TypeScript criamos um array de chaves para evitar problemas de mutação durante a iteração
        const chavesEfeitos = Object.keys(this.efeitos);
        for (const efeito of chavesEfeitos) {
            if (EFEITOS_DANO_RECEBIDO[efeito] && this.efeitos[efeito] > 0) {
                danoReal = EFEITOS_DANO_RECEBIDO[efeito](danoReal);
                this.efeitos[efeito] -= 1;
                if (this.efeitos[efeito] <= 0) {
                    efeitosParaRemover.push(efeito);
                }
            }
        }

        for (const efeito of efeitosParaRemover) {
            delete this.efeitos[efeito];
        }

        this.hp_atual -= danoReal;
        if (this.hp_atual < 0) {
            this.hp_atual = 0;
        }
        
        return danoReal;
    }

    ganharXp(quantidade: number): void {
        this.xp += quantidade;
        let xpNecessario = 30 + (this.nivel - 1) * 60;
        
        while (this.xp >= xpNecessario) {
            this.xp -= xpNecessario;
            this.nivel += 1;
            this.hp_max += Math.floor(12 * calcularMultiplicadorRebirth(this.rebirths));
            this.hp_atual = this.hp_max;
            
            if (GALOS_DB[this.nome]) {
                const skillsNoNovoNivel = [];
                for (const [lvlStr, skill] of Object.entries(GALOS_DB[this.nome].skills)) {
                    if (parseInt(lvlStr, 10) === this.nivel) {
                        skillsNoNovoNivel.push(skill);
                    }
                }
                
                if (skillsNoNovoNivel.length > 0) {
                    this.skills_equipadas = [...skillsNoNovoNivel, ...this.skills_equipadas];
                    this.skills_equipadas = this.skills_equipadas.slice(0, 5);
                }
            }
            
            while (this.skills_equipadas.length < 5) {
                this.skills_equipadas.push(null);
            }

            xpNecessario = 30 + (this.nivel - 1) * 60;
        }
    }

    toDict(): any {
        return {
            nome: this.nome,
            tipo: this.tipo,
            hp_max: this.hp_max,
            caminho_imagem: this.caminho_imagem,
            nivel: this.nivel,
            xp: this.xp,
            skills_equipadas: this.skills_equipadas,
            efeitos: this.efeitos,
            rebirths: this.rebirths
        };
    }

    static fromDict(data: any): Galo {
        const skills_equipadas = Array.isArray(data.skills_equipadas) ? [...data.skills_equipadas] : [];
        while (skills_equipadas.length < 5) {
            skills_equipadas.push(null);
        }
        
        return new Galo(
            data.nome,
            data.hp_max,
            data.caminho_imagem,
            data.nivel ?? 1,
            data.xp ?? 0,
            data.tipo ?? "Normal",
            skills_equipadas,
            data.efeitos ? { ...data.efeitos } : {},
            data.rebirths ?? 0
        );
    }
}

File: rinha-offline/codigo_antigo/logic/galo.py
import random
from data.galos_db import GALOS_DB
from logic.efeitos import EFEITOS_INICIO_TURNO, EFEITOS_DANO_RECEBIDO # Importação do motor de efeitos

class Galo:
    def __init__(self, nome, hp_max, caminho_imagem, nivel=1, xp=0, tipo="Normal", skills_equipadas=None, efeitos=None): 
        self.nome = nome
        self.tipo = tipo
        self.nivel = nivel
        self.xp = xp
        self.hp_max = hp_max
        self.hp_atual = hp_max
        self.caminho_imagem = caminho_imagem
        self.skills_equipadas = skills_equipadas if skills_equipadas is not None else [] 
        self.efeitos = efeitos if efeitos is not None else {} 

    def obter_skills_desbloqueadas(self):
        if self.nome not in GALOS_DB:
            return []
        skills_db = GALOS_DB[self.nome]["skills"]
        desbloqueadas = []
        for lvl, skill in skills_db.items():
            if lvl <= self.nivel:
                desbloqueadas.append((lvl, skill))
        return desbloqueadas

    def equipar_skills_bot(self):
        desbloqueadas = self.obter_skills_desbloqueadas()
        desbloqueadas.sort(key=lambda x: x[0], reverse=True)
        self.skills_equipadas = [skill for lvl, skill in desbloqueadas[:5]]
        while len(self.skills_equipadas) < 5:
            self.skills_equipadas.append(None)

    def atacar(self):
        skills_validas = [s for s in self.skills_equipadas if s is not None]
        if not skills_validas:
            self.equipar_skills_bot()
            skills_validas = [s for s in self.skills_equipadas if s is not None]
            
        if not skills_validas:
            return "Ataque Básico", 10, None

        skill = random.choice(skills_validas)
        dano_total = random.randint(skill["min"], skill["max"])
        
        dados_efeito = None
        if "efeito" in skill:
            dados_efeito = {
                "nome": skill["efeito"],
                "chance": skill.get("chance", 100),
                "turnos": skill.get("turnos", 1)
            }
            
        return skill["nome"], dano_total, dados_efeito

    def aplicar_efeito(self, dados_efeito):
        if not dados_efeito:
            return False
            
        if random.randint(1, 100) <= dados_efeito["chance"]:
            self.efeitos[dados_efeito["nome"]] = dados_efeito["turnos"]
            return True
            
        return False

    def processar_efeitos_inicio_turno(self):
        mensagens = []
        pode_atacar = True
        efeitos_para_remover = []

        for efeito, turnos in self.efeitos.items():
            if turnos > 0:
                if efeito in EFEITOS_INICIO_TURNO: # Delega o processamento da mecânica para o motor
                    msg, permite_atacar = EFEITOS_INICIO_TURNO[efeito](self, efeito)
                    if msg: mensagens.append(msg)
                    if not permite_atacar: pode_atacar = False

                self.efeitos[efeito] -= 1
                if self.efeitos[efeito] <= 0:
                    efeitos_para_remover.append(efeito)

        for efeito in efeitos_para_remover:
            del self.efeitos[efeito]

        if self.hp_atual < 0:
            self.hp_atual = 0

        return mensagens, pode_atacar

    def sofrer_dano(self, dano_recebido):
        dano_real = max(1, dano_recebido)
        efeitos_para_remover = []
        
        for efeito in list(self.efeitos.keys()): # Itera iterando cópia das chaves para processar mitigações reativas
            if efeito in EFEITOS_DANO_RECEBIDO and self.efeitos[efeito] > 0:
                dano_real = EFEITOS_DANO_RECEBIDO[efeito](dano_real)
                self.efeitos[efeito] -= 1
                if self.efeitos[efeito] <= 0:
                    efeitos_para_remover.append(efeito)
                    
        for efeito in efeitos_para_remover:
            del self.efeitos[efeito]

        self.hp_atual -= dano_real
        if self.hp_atual < 0:
            self.hp_atual = 0
        return dano_real

    def ganhar_xp(self, quantidade):
        self.xp += quantidade
        xp_necessario = 30 + (self.nivel - 1) * 60
        
        while self.xp >= xp_necessario:
            self.xp -= xp_necessario
            self.nivel += 1
            self.hp_max += 12
            self.hp_atual = self.hp_max
            xp_necessario = 30 + (self.nivel - 1) * 60

    def to_dict(self):
        return {
            "nome": self.nome,
            "tipo": self.tipo,
            "hp_max": self.hp_max,
            "caminho_imagem": self.caminho_imagem,
            "nivel": self.nivel,
            "xp": self.xp,
            "skills_equipadas": self.skills_equipadas,
            "efeitos": self.efeitos
        }

    @classmethod
    def from_dict(cls, data):
        skills_equipadas = data.get("skills_equipadas", [])
        while len(skills_equipadas) < 5:
            skills_equipadas.append(None)
            
        return cls(
            data["nome"], 
            data["hp_max"], 
            data["caminho_imagem"], 
            data.get("nivel", 1), 
            data.get("xp", 0),
            data.get("tipo", "Normal"),
            skills_equipadas,
            data.get("efeitos", {})
        )